### ISSUE TITLE
Missing re-ordering of `exp_avg` in SOAP

### ISSUE THREAD

**Issue by francois-rozet** [HomebrewML/HeavyBall#42]
In the original implementation of SOAP, `exp_avg` is computed outside of the eigenbasis and projected at every update. In this implementation, it is computed directly in the eigenbasis, which is slightly more efficient. However, this means that

a. as the eigenbasis changes, the moving average becomes inaccurate
b. the elements in `exp_avg` should be re-ordered in the same manner as `exp_avg_sq` (which it is not).

https://github.com/HomebrewML/HeavyBall/blob/f2cb0abd5e5675215248b7e08e114bb4c83a8faa/heavyball/utils.py#L388

This can be fixed by projecting `exp_avg` back before the precond update.

**Comment by ClashLuke** [comment]
You're right, there's definitely still at least some difference
![Image](https://github.com/user-attachments/assets/eee42f38-66fe-4d27-bed2-a12535f1085d)

**Comment by ClashLuke** [comment]
With lr=1e-5, we converge better. With lr=1e-3, baseline SOAP. So, it might be something that increases the effective learning rate.

Updated code:
```
import heavyball
import torch
import torch.nn as nn
import matplotlib.pyplot as plt
import numpy as np
import traceback

from heavyball import ForeachSOAP
from soap import SOAP
from tqdm import trange

heavyball.utils.compile_mode = None

# Dictionary to store losses for each optimizer and datatype
losses = {}

# List of datatypes to test
dtypes = [torch.float32, torch.float64]

lr = 1e-3
eps = 1e-10

for dtype in dtypes:
    for name in ("soap", "foreach-soap"):
        # Create unique key for each optimizer-datatype combination
        key = f"{name}_{str(dtype).split('.')[-1]}"
        losses[key] = []
        
        try:
            _ = torch.manual_seed(0)
            x = torch.randn(1, 3, 16, 16).to(dtype)

            net = nn.Sequential(
                nn.Conv2d(3, 64, kernel_size=3, padding=1),
                nn.ReLU(),
                nn.Conv2d(64, 3, kernel_size=3, padding=1),
            ).to(dtype)

            if name == "soap":
                optimizer = SOAP(
                    net.parameters(),
                    lr=lr,
                    betas=[0.9, 0.999],
                    eps=eps,
                    weight_decay=0.0,
                    precondition_frequency=16,
                    max_precond_dim=4096,
                    merge_dims=False,
                )
            else:
                optimizer = ForeachSOAP(
                    net.parameters(),
                    lr=lr,
                    storage_dtype=str(dtype).split('.')[-1],
                    betas=[0.9, 0.999],
                    shampoo_beta=0.999,
                    eps=eps,
                    weight_decay=0.0,
                    precondition_frequency=16,
                    max_precond_dim=4096,
                    merge_dims=False,
                )

            # Training loop with loss tracking
            for _ in (bar := trange(10000)):
                y = net(x)
                l = (x - y).square().mean()
                l.backward()

                optimizer.step()
                optimizer.zero_grad()

                # Store the loss value
                losses[key].append(l.item())
                bar.set_postfix(loss=l.item())

        except Exception as e:
            print(f"Error occurred for {key}:")
            traceback.print_exc()
            print(f"Skipping {key} and continuing with next combination\n")
            # Clear any partial results for this failed combination
            losses[key] = []

# Create the convergence plot
plt.figure(figsize=(12, 7))
for key, loss_values in losses.items():
    if loss_values:  # Only plot if we have data
        plt.plot(np.arange(len(loss_values)), loss_values, label=key)

plt.yscale('log')  # Set logarithmic scale on y-axis
plt.xlabel('Iteration')
plt.ylabel('Loss')
plt.title('Convergence Comparison: SOAP vs Foreach-SOAP across Data Types')
plt.legend()
plt.grid(True, which="both", ls="-", alpha=0.2)
plt.show()
```

![Image](https://github.com/user-attachments/assets/5cb3e187-775f-40ab-acbc-175ed5d42707)

**Comment by francois-rozet** [comment]
I have a clean implementation of SOAP with the corrections here:

https://gist.github.com/francois-rozet/b735179003c9d4eea2a942f096ceae09

In my expriments it does strictly better than the current ForeachSOAP.

**Comment by francois-rozet** [comment]
In my code, `exp_avg` is projected back and forth at every `Q` update, which is not that frequent. In fact, I don't think that errors can accumulate in `exp_avg` fast enough as $\beta_1$ is usually around 0.9.

**Comment by MJK12341234** [comment]
This is implemented in the original SOAP code (https://github.com/nikhilvyas/SOAP/blob/a1e553530fde97d0e6b307d7c82ac6d38b072340/soap.py#L268-L274). We project back and forth exponential average every time update_preconditioner is called.

**Comment by francois-rozet** [comment]
> This is implemented in the original SOAP code (https://github.com/nikhilvyas/SOAP/blob/a1e553530fde97d0e6b307d7c82ac6d38b072340/soap.py#L268-L274). We project back and forth exponential average every time update_preconditioner is called. 

This is wasteful. You only need to project when Q is updated.

**Comment by ClashLuke** [comment]
I'll have to hand over the torch for now. The SOAP part seems to be correct, the key difference is in our Adam implementation 4d32368e2917936b6199f4814dd600c07af8e597

```
import heavyball
import torch
import torch.nn as nn
import matplotlib.pyplot as plt
import numpy as np
import traceback

from heavyball import ForeachSOAP
from soap import SOAP
from tqdm import trange
from torch.backends import cudnn, opt_einsum

heavyball.utils.compile_mode = None

opt_einsum.enabled = True
opt_einsum.strategy = "dp"


# Dictionary to store losses for each optimizer and datatype
losses = {}

# List of datatypes to test
dtypes = [torch.float64]

lr = 1e-3
eps = 1e-12
freq = 1

rand_mats = {}

dtype = torch.float64

for eps in (1e-6, 1e-10):
    for name in ("foreach-soap", 'soap'):
        # Create unique key for each optimizer-datatype combination
        key = f"{name}_{str(dtype).split('.')[-1]}_{eps}"
        losses[key] = []
        print('')
        try:
            _ = torch.manual_seed(0)
            x = torch.randn(16, 3, 16, 16).to(dtype).cuda()

            net = nn.Sequential(
                nn.Conv2d(3, 64, kernel_size=3, padding=1),
                #nn.ReLU(),
                nn.Conv2d(64, 3, kernel_size=3, padding=1),
            ).to(dtype).cuda()

            if name == "soap":
                optimizer = SOAP(
                    net.parameters(),
                    lr=lr,
                    betas=[0.9, 0.999, 0.999],
                    eps=eps,
                    weight_decay=0.0,
                    precondition_frequency=freq,
                    max_precond_size=4096,
                    merge_dims=False,
                )
            else:
                optimizer = ForeachSOAP(
                    net.parameters(),
                    lr=lr,
                    storage_dtype=str(dtype).split('.')[-1],
                    betas=[0.9, 0.999],
                    shampoo_beta=0.999,
                    eps=eps,
                    weight_decay=0.0,
                    precondition_frequency=freq,
                    max_precond_dim=4096,
                    merge_dims=False,
                )


            # Training loop with loss tracking
            for i in (bar := range(20)):
                y = net(x)
                l = (x - y).square().mean()
                l.backward()

                optimizer.step()
                optimizer.zero_grad()

                with torch.no_grad():
                    # Store the loss value
                    losses[key].append(l.detach())
                    #bar.set_postfix(loss=l.item())

        except Exception as e:
            print(f"Error occurred for {key}:")
            traceback.print_exc()
            print(f"Skipping {key} and continuing with next combination\n")
            # Clear any partial results for this failed combination
            losses[key] = []

# Create the convergence plot
plt.figure(figsize=(12, 7))
for key, loss_values in losses.items():
    if loss_values:  # Only plot if we have data
        plt.plot(np.arange(len(loss_values)), [i.item() for i in loss_values], label=key)

plt.yscale('log')  # Set logarithmic scale on y-axis
plt.xlabel('Iteration')
plt.ylabel('Loss')
plt.title('Convergence Comparison: SOAP vs Foreach-SOAP across Data Types')
plt.legend()
plt.grid(True, which="both", ls="-", alpha=0.2)
plt.show()
```

**Comment by francois-rozet** [comment]
While you are modifying your implementation, you might want to look at my PR (or https://gist.github.com/francois-rozet/b735179003c9d4eea2a942f096ceae09), which features other improvements (better dim merging, debiased shampoo's beta, Q update that does not require reordering exp_avg_sq, overall clean up).

**Comment by ClashLuke** [comment]
Starting 3f73d24f0171245dad17529615e4fa55c1583502, SOAP works again.\
I'll push a new version (`1.6.0`), rerun the benchmarks, and share the news in the next few days. (SOAP shouldn't suffer from my mistake.)\
Thank you @francois-rozet for catching this and providing the reference implementation (which turned out to be 3x faster than my approach), and @nikhilvyas for clarifying the original design intent.

FP64:

![Image](https://github.com/user-attachments/assets/6be7122a-2934-4233-af29-48a45e5574a2)

![Image](https://github.com/user-attachments/assets/a08e2571-88f7-4ea8-98f4-c82b2af914e9)

FP32:

![Image](https://github.com/user-attachments/assets/c2c3eed2-a6eb-4c57-aee7-808b462a5fa3)

![Image](https://github.com/user-attachments/assets/afc8b83e-f3ed-4287-b94e-9837545db11d)
